os/services/orchestra/orchestra-rule-eb-per-time-source.c

```diff
@@ -74,47 +91,51 @@ select_packet(uint16_t *slotframe, uint16_t *timeslot, uint16_t *channel_offset)
 static void
 new_time_source(const struct tsch_neighbor *old, const struct tsch_neighbor *new)
 {
-  uint16_t old_ts = old != NULL ? get_node_timeslot(tsch_queue_get_nbr_address(old)) : 0xffff;
-  uint16_t new_ts = new != NULL ? get_node_timeslot(tsch_queue_get_nbr_address(new)) : 0xffff;
+  uint16_t old_ts = 0xffff;
+  uint16_t new_ts = 0xffff;
+  uint16_t old_channel_offset = 0xffff;
+  uint16_t new_channel_offset = 0xffff;
+
+  if(old != NULL) {
+    const linkaddr_t *addr = tsch_queue_get_nbr_address(old);
+    old_ts = get_node_timeslot(addr);
+    old_channel_offset = get_node_channel_offset(addr);
+  }
 
-  if(new_ts == old_ts) {
+  if(new != NULL) {
+    const linkaddr_t *addr = tsch_queue_get_nbr_address(new);
+    new_ts = get_node_timeslot(addr);
+    new_channel_offset = get_node_channel_offset(addr);
+  }
+
+  if(new_ts == old_ts && old_channel_offset == new_channel_offset) {
     return;
   }
 
   if(old_ts != 0xffff) {
     /* Stop listening to the old time source's EBs */
-    if(old_ts == get_node_timeslot(&linkaddr_node_addr)) {
-      /* This was the same timeslot as slot. Reset original link options */
-      tsch_schedule_add_link(sf_eb, LINK_OPTION_TX, LINK_TYPE_ADVERTISING_ONLY,
-        &tsch_broadcast_address, old_ts, ORCHESTRA_EB_CHANNEL_OFFSET, 1);
-    } else {
-      /* Remove slot */
-      tsch_schedule_remove_link_by_timeslot(sf_eb, old_ts, ORCHESTRA_EB_CHANNEL_OFFSET);
-    }
+    tsch_schedule_remove_link_by_timeslot(sf_eb, old_ts, old_channel_offset);
   }
   if(new_ts != 0xffff) {
-    uint8_t link_options = LINK_OPTION_RX;
-    if(new_ts == get_node_timeslot(&linkaddr_node_addr)) {
-      /* This is also our timeslot, add necessary flags */
-      link_options |= LINK_OPTION_TX;
-    }
     /* Listen to the time source's EBs */
-    tsch_schedule_add_link(sf_eb, link_options, LINK_TYPE_ADVERTISING_ONLY,
-      &tsch_broadcast_address, new_ts, ORCHESTRA_EB_CHANNEL_OFFSET, 1);
+    tsch_schedule_add_link(sf_eb, LINK_OPTION_RX, LINK_TYPE_ADVERTISING_ONLY,
+      &tsch_broadcast_address, new_ts, new_channel_offset, 0);
```

Comment:
Old code had `do_remove` parameter of `tsch_schedule_add_link` set to 1. New code has it set to 0. The intention behind the PR was to simplify the logic in the file by making this change. Do you think there is a possible problem? I will look this into depth if you still think its suspicious.

Reply:
When seeing this the first time I was thinking that `tsch_schedule_add_link()` with `do_remove` set to 0 would simply update an existing link with new options. But when inspecting the code it seems like it rather adds a second link in addition to the existing one. And with that in mind I do not see this working. Am I reading it wrong?

Reply:
If the timeslots match, then there will be two links at the same time offset. One with Tx flag, for transmissions from the local node, one with Rx flag for receptions from the parent. The channel offsets may be same of different. Why do you think it is the issue?

Reply:
Ah, my bad, I was misunderstanding how links with identical ts/ch was handled.

Let's see if my last open item is valid:
Is there anything stopping `tsch_schedule_remove_link_by_timeslot()` at line 119 from inadvertently deleting both the TX and RX link? E.g. if we lose the time-source, we call `new_time_source(old_addr, NULL)`, which would call the mentioned function and potentially remove our TX as well(?).

Reply:
Yes, you're right that there is a problem! If both channel offsets match, it will remove both.
I changed the code to fix this, thanks for spotting it.

Reply:
:+1: